internal/ratelimit/http.go

```diff
@@ -0,0 +1,127 @@
+	"github.com/observatorium/observatorium/internal/ratelimit/gubernator"
+)
+
+const (
+	requestName    = "observatorium:tenant_per_endpoint"
+	requestTimeout = time.Millisecond * 500
+)
+
+// Config configures a rate limiter per endpoint, per tenant.
+type Config struct {
+	Tenant  string
+	Matcher *regexp.Regexp
+	Limit   int
+	Window  time.Duration
+}
+
+type middleware struct {
+	matcher *regexp.Regexp
+	handler internal.Middleware
+}
+
+// WithLocalRateLimiter returns a middleware that controls amount of requests per tenant using in-memory store.
+func WithLocalRateLimiter(configs ...Config) internal.Middleware {
+	middlewares := make(map[string][]middleware)
+	for _, c := range configs {
+		middlewares[c.Tenant] = append(middlewares[c.Tenant], middleware{c.Matcher, httprate.NewRateLimiter(
+			c.Limit,
+			c.Window,
+			nil,
+		).Handler})
+	}
+
+	return combine(middlewares)
+}
+
+// WithSharedRateLimiter returns a middleware that controls amount of requests per tenant using external service.
+func WithSharedRateLimiter(client gubernator.V1Client, configs ...Config) internal.Middleware {
+	middlewares := make(map[string][]middleware)
+	for _, c := range configs {
+		middlewares[c.Tenant] = append(middlewares[c.Tenant], middleware{c.Matcher, rateLimiter{client, &gubernator.RateLimitReq{
+			Name:      requestName,
+			UniqueKey: fmt.Sprintf("%s:%s", c.Tenant, c.Matcher.String()),
+			Hits:      1,
+			Limit:     int64(c.Limit),
+			Duration:  c.Window.Microseconds(),
+			Algorithm: gubernator.Algorithm_LEAKY_BUCKET,
+		}}.Handler})
+	}
+
+	return combine(middlewares)
+}
+
+func combine(middlewares map[string][]middleware) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			tenant, ok := authentication.GetTenant(r.Context())
+			if !ok {
+				// This shouldn't have happened.
+				http.Error(w, "error finding tenant", http.StatusUnauthorized)
+				return
+			}
+
+			tms, ok := middlewares[tenant]
+			if !ok {
+				// No rate limits configured for this tenant.
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			for _, m := range tms {
+				if m.matcher.MatchString(r.URL.Path) {
+					m.handler(next).ServeHTTP(w, r)
+					return
+				}
+			}
+
+			// No rate limits configured for this endpoint.
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+type rateLimiter struct {
+	client gubernator.V1Client
+
+	req *gubernator.RateLimitReq
+}
+
+func (l rateLimiter) Handler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+		defer cancel()
+
+		resp, err := l.client.GetRateLimits(ctx, &gubernator.GetRateLimitsReq{
+			Requests: []*gubernator.RateLimitReq{l.req},
+		})
+		if err != nil {
+			// TODO(kakkoyun): Should we add metric and fail gracefully gracefully?
```

Comment:
I guess an error is only returned if the underlying service has an actual problem or there was some network issue, right? An expected error from the services should be part of the resp payload?!
In that case, I'd say failing hard is the preferred way.

Reply:
This is an important decision to make. We are making rate-limiting a hard dependency on request life time. I think similar to auth it should be blocking, but it will be painful if this will become flaky. I hope metric is there with grpc client metric middleware? (: